vpr/src/route/rr_graph_reader.cpp

```diff
@@ -47,809 +35,751 @@
-void process_seg_id(pugi::xml_node parent, const pugiutil::loc_data & loc_data);
-void set_cost_indices(pugi::xml_node parent, const pugiutil::loc_data& loc_data,
-        const bool is_global_graph, const int num_seg_types);
-
-/************************ Subroutine definitions ****************************/
+void set_cost_indices(bool is_global_graph, const int num_seg_types);
+void process_seg_id();
+
+/********************* Global variables for this module *******************/
+/* libxml2 interface */
+static xmlSAXHandler sax_handler = {};
+
+/* We need to build the current <meta> node piece by piece here. For edge metadata,
+ information about the current edge is also required. */
+static std::string current_meta_name;
+static std::string current_meta_value;
+static enum {NODE, EDGE} current_meta_place;
+static struct {int src_id; int sink_id; int switch_id;} current_edge;
+
+/* The <edge> handler(consume_node()) tracks the most frequently appearing switch that
+ * connects a CHANX/CHANY node to a IPIN node. It is then assigned to *wire_to_rr_ipin_switch. */
+int *count_for_wire_to_ipin_switches = NULL;
+struct {int id; int count;} most_frequent_switch = {-1, 0};
+
+/* current_node_id is kept because we need to build a map from node IDs to segment IDs
+ * and t_rr_node does not hold IDs.
+ * !! Also note that it's assumed that nodes appear in ID order and no ID gaps occur. */
+static int current_node_id;
+
+/* Pointers to arguments to load_rr_file: we need to access them from SAX handlers. */
+static t_chan_width *chan_width;
+static const DeviceGrid *g_grid;
+
+/* Map tag names to ints. This helps us build a lookup table instead of a hashmap,
+ * and it also enables us to use an array of ints instead of a std::stack of std::strings as the parser stack. */
+enum tag_type : int {
+	T_RR_GRAPH = 1,
+	T_CHANNELS,
+	T_SWITCHES,
+	T_SEGMENTS,
+	T_BLOCK_TYPES,
+	T_GRID,
+	T_RR_NODES,
+	T_RR_EDGES,
+	T_CHANNEL,
+	T_X_LIST,
+	T_Y_LIST,
+	T_SWITCH,
+	T_TIMING,
+	T_SIZING,
+	T_SEGMENT,
+	T_BLOCK_TYPE,
+	T_PIN_CLASS,
+	T_PIN,
+	T_GRID_LOC,
+	T_NODE,
+	T_EDGE,
+	T_METADATA,
+	T_META,
+	T_LOC,
+};
+static std::unordered_map<std::string, tag_type> tag_map = {
+	{"rr_graph", T_RR_GRAPH},
+	{"channels", T_CHANNELS},
+	{"switches", T_SWITCHES},
+	{"segments", T_SEGMENTS},
+	{"block_types", T_BLOCK_TYPES},
+	{"grid", T_GRID},
+	{"rr_nodes", T_RR_NODES},
+	{"rr_edges", T_RR_EDGES},
+	{"channel", T_CHANNEL},
+	{"x_list", T_X_LIST},
+	{"y_list", T_Y_LIST},
+	{"switch", T_SWITCH},
+	{"timing", T_TIMING},
+	{"sizing", T_SIZING},
+	{"segment", T_SEGMENT},
+	{"block_type", T_BLOCK_TYPE},
+	{"pin_class", T_PIN_CLASS},
+	{"pin", T_PIN},
+	{"grid_loc", T_GRID_LOC},
+	{"node", T_NODE},
+	{"edge", T_EDGE},
+	{"metadata", T_METADATA},
+	{"meta", T_META},
+	{"loc", T_LOC},
+};
+
+/* The parser stack to keep the current path of tags.
+ * It can have a static size, because there is no recursive nesting in the file format. */
+static tag_type parser_stack[64];
```

Comment:
Why 64?

Reply:
it's arbitrary. the file should have 6 or 7 nested tags maximum, since there is no recursive rule.